// @atlaskit/select/dist/esm/PopupSelect/PopupSelect.js
import _extends from "@babel/runtime/helpers/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/objectWithoutProperties";
import _classCallCheck from "@babel/runtime/helpers/classCallCheck";
import _createClass from "@babel/runtime/helpers/createClass";
import _assertThisInitialized from "@babel/runtime/helpers/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/inherits";
import _possibleConstructorReturn from "@babel/runtime/helpers/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/getPrototypeOf";
import _defineProperty from "@babel/runtime/helpers/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) { symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); } keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

import React, { PureComponent } from 'react';
import { createPortal } from 'react-dom';
import Select, { mergeStyles } from 'react-select';
import { uid } from 'react-uid';
import createFocusTrap from 'focus-trap';
import { Manager, Reference, Popper } from 'react-popper';
import NodeResolver from 'react-node-resolver';
import shallowEqualObjects from 'shallow-equal/objects';
import { N80 } from '@atlaskit/theme/colors';
import { token } from '@atlaskit/tokens';
import { MenuDialog, DummyControl, defaultComponents } from './components';
import baseStyles from '../styles';

/** Are we rendering on the client or server? */
var canUseDOM = function canUseDOM() {
  return Boolean(typeof window !== 'undefined' && window.document && window.document.createElement);
}; // ==============================
// Types
// ==============================


// ==============================
// Class
// ==============================
var defaultPopperProps = {
  modifiers: [{
    name: 'offset',
    options: {
      offset: [0, 8]
    }
  }, {
    name: 'preventOverflow',
    enabled: true,
    options: {
      padding: 5,
      boundary: 'clippingParents',
      altAxis: true,
      altBoundary: true
    }
  }],
  placement: 'bottom-start'
};

var isEmpty = function isEmpty(obj) {
  return Object.keys(obj).length === 0;
};

var PopupSelect = /*#__PURE__*/function (_PureComponent) {
  _inherits(PopupSelect, _PureComponent);

  var _super = _createSuper(PopupSelect);

  function PopupSelect() {
    var _this;

    _classCallCheck(this, PopupSelect);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _super.call.apply(_super, [this].concat(args));

    _defineProperty(_assertThisInitialized(_this), "focusTrap", null);

    _defineProperty(_assertThisInitialized(_this), "menuRef", null);

    _defineProperty(_assertThisInitialized(_this), "selectRef", null);

    _defineProperty(_assertThisInitialized(_this), "targetRef", null);

    _defineProperty(_assertThisInitialized(_this), "defaultStyles", mergeStyles(baseStyles(_this.props.validationState, _this.props.spacing === 'compact'), {
      groupHeading: function groupHeading(provided) {
        return _objectSpread(_objectSpread({}, provided), {}, {
          color: token('color.text.lowEmphasis', N80)
        });
      }
    }));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isOpen: false,
      mergedComponents: defaultComponents,
      mergedPopperProps: defaultPopperProps
    });

    _defineProperty(_assertThisInitialized(_this), "popperWrapperId", "".concat(uid({
      options: _this.props.options
    }), "-popup-select"));

    _defineProperty(_assertThisInitialized(_this), "handleKeyDown", function (event) {
      switch (event.key) {
        case 'Escape':
        case 'Esc':
          _this.close();

          break;

        default:
      }

      if (_this.props.handleKeyDown) {
        _this.props.handleKeyDown(event);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleClick", function (_ref) {
      var target = _ref.target;
      var isOpen = _this.state.isOpen; // appease flow

      if (!(target instanceof Element)) {
        return;
      } // NOTE: Why not use the <Blanket /> component to close?
      // We don't want to interupt the user's flow. Taking this approach allows
      // user to click "through" to other elements and close the popout.


      if (isOpen && _this.menuRef && !_this.menuRef.contains(target)) {
        _this.close();
      } // open on target click -- we can't trust consumers to spread the onClick
      // property to the target


      if (!isOpen && _this.targetRef && _this.targetRef.contains(target)) {
        _this.open();
      }
    });

    _defineProperty(_assertThisInitialized(_this), "handleSelectChange", function (value, actionMeta) {
      var _this$props = _this.props,
          closeMenuOnSelect = _this$props.closeMenuOnSelect,
          onChange = _this$props.onChange;

      if (closeMenuOnSelect && actionMeta.action !== 'clear') {
        _this.close();
      }

      if (onChange) {
        onChange(value, actionMeta);
      }
    });

    _defineProperty(_assertThisInitialized(_this), "open", function () {
      var onOpen = _this.props.onOpen;

      if (onOpen) {
        onOpen();
      }

      _this.setState({
        isOpen: true
      }, _this.initialiseFocusTrap);

      if (_this.selectRef) {
        _this.selectRef.select.openMenu('first'); // HACK

      }

      if (typeof window === 'undefined') {
        return;
      }

      window.addEventListener('keydown', _this.handleKeyDown, true);
    });

    _defineProperty(_assertThisInitialized(_this), "initialiseFocusTrap", function () {
      if (!_this.menuRef) {
        return;
      }

      var trapConfig = {
        clickOutsideDeactivates: true,
        escapeDeactivates: true,
        fallbackFocus: _this.menuRef,
        returnFocusOnDeactivate: true
      };
      _this.focusTrap = createFocusTrap(_this.menuRef, trapConfig); // allow time for the HTMLElement to render

      setTimeout(function () {
        return _this.focusTrap.activate();
      }, 1);
    });

    _defineProperty(_assertThisInitialized(_this), "close", function () {
      var onClose = _this.props.onClose;

      if (onClose) {
        onClose();
      }

      _this.setState({
        isOpen: false
      });

      if (_this.focusTrap) {
        _this.focusTrap.deactivate();
      }

      if (typeof window === 'undefined') {
        return;
      }

      window.removeEventListener('keydown', _this.handleKeyDown, true);
    });

    _defineProperty(_assertThisInitialized(_this), "resolveTargetRef", function (popperRef) {
      return function (ref) {
        // avoid thrashing fn calls
        if (!_this.targetRef && popperRef && ref) {
          _this.targetRef = ref;

          if (typeof popperRef === 'function') {
            popperRef(ref);
          } else {
            popperRef.current = ref;
          }
        }
      };
    });

    _defineProperty(_assertThisInitialized(_this), "resolveMenuRef", function (popperRef) {
      return function (ref) {
        _this.menuRef = ref;

        if (typeof popperRef === 'function') {
          popperRef(ref);
        } else {
          popperRef.current = ref;
        }
      };
    });

    _defineProperty(_assertThisInitialized(_this), "getSelectRef", function (ref) {
      _this.selectRef = ref;
    });

    _defineProperty(_assertThisInitialized(_this), "getItemCount", function () {
      var options = _this.props.options;
      var count = 0;
      options.forEach(function (groupOrOption) {
        if (groupOrOption.options) {
          groupOrOption.options.forEach(function () {
            return count++;
          });
        } else {
          count++;
        }
      });
      return count;
    });

    _defineProperty(_assertThisInitialized(_this), "getMaxHeight", function () {
      var maxMenuHeight = _this.props.maxMenuHeight;

      if (!_this.selectRef) {
        return maxMenuHeight;
      } // subtract the control height to maintain consistency


      var showSearchControl = _this.showSearchControl();

      var controlRef = _this.selectRef.select.controlRef; // @ts-ignore React-select provides incomplete types for controlRef

      var offsetHeight = showSearchControl ? controlRef.offsetHeight : 0;
      var maxHeight = maxMenuHeight - offsetHeight;
      return maxHeight;
    });

    _defineProperty(_assertThisInitialized(_this), "showSearchControl", function () {
      var searchThreshold = _this.props.searchThreshold;
      return _this.getItemCount() > searchThreshold;
    });

    _defineProperty(_assertThisInitialized(_this), "renderSelect", function () {
      var _this$props2 = _this.props,
          footer = _this$props2.footer,
          maxMenuWidth = _this$props2.maxMenuWidth,
          minMenuWidth = _this$props2.minMenuWidth,
          target = _this$props2.target,
          props = _objectWithoutProperties(_this$props2, ["footer", "maxMenuWidth", "minMenuWidth", "target"]);

      var _this$state = _this.state,
          isOpen = _this$state.isOpen,
          mergedComponents = _this$state.mergedComponents,
          mergedPopperProps = _this$state.mergedPopperProps;

      var showSearchControl = _this.showSearchControl();

      var portalDestination = canUseDOM() ? document.body : null;

      var components = _objectSpread(_objectSpread({}, mergedComponents), {}, {
        Control: showSearchControl ? mergedComponents.Control : DummyControl
      });

      if (!portalDestination || !isOpen) {
        return null;
      }

      var popper = /*#__PURE__*/React.createElement(Popper, mergedPopperProps, function (_ref2) {
        var placement = _ref2.placement,
            ref = _ref2.ref,
            style = _ref2.style;
        return /*#__PURE__*/React.createElement(NodeResolver, {
          innerRef: _this.resolveMenuRef(ref)
        }, /*#__PURE__*/React.createElement(MenuDialog, {
          style: style,
          "data-placement": placement,
          minWidth: minMenuWidth,
          maxWidth: maxMenuWidth,
          id: _this.popperWrapperId
        }, /*#__PURE__*/React.createElement(Select, _extends({
          backspaceRemovesValue: false,
          controlShouldRenderValue: false,
          isClearable: false,
          tabSelectsValue: false,
          menuIsOpen: true,
          ref: _this.getSelectRef
        }, props, {
          isSearchable: showSearchControl,
          styles: _objectSpread(_objectSpread({}, _this.defaultStyles), props.styles),
          maxMenuHeight: _this.getMaxHeight(),
          components: components,
          onChange: _this.handleSelectChange
        })), footer));
      });
      return mergedPopperProps.strategy === 'fixed' ? popper : /*#__PURE__*/createPortal(popper, portalDestination);
    });

    return _this;
  }

  _createClass(PopupSelect, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      if (typeof window === 'undefined') {
        return;
      }

      window.addEventListener('click', this.handleClick, true);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      if (typeof window === 'undefined') {
        return;
      }

      window.removeEventListener('click', this.handleClick, true);
      window.removeEventListener('keydown', this.handleKeyDown, true);
    } // Event Handlers
    // ==============================

  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      var target = this.props.target;
      var isOpen = this.state.isOpen;
      return /*#__PURE__*/React.createElement(Manager, null, /*#__PURE__*/React.createElement(Reference, null, function (_ref3) {
        var ref = _ref3.ref;
        return target && target({
          isOpen: isOpen,
          ref: _this2.resolveTargetRef(ref),
          'aria-haspopup': 'true',
          'aria-expanded': isOpen,
          'aria-controls': isOpen ? _this2.popperWrapperId : undefined
        });
      }), this.renderSelect());
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props, state) {
      var newState = {}; // Merge consumer and default popper props

      var mergedPopperProps = _objectSpread(_objectSpread({}, defaultPopperProps), props.popperProps);

      if (!shallowEqualObjects(mergedPopperProps, state.mergedPopperProps)) {
        newState.mergedPopperProps = mergedPopperProps;
      } // Merge consumer and default components


      var mergedComponents = _objectSpread(_objectSpread({}, defaultComponents), props.components);

      if (!shallowEqualObjects(mergedComponents, state.mergedComponents)) {
        newState.mergedComponents = mergedComponents;
      }

      if (!isEmpty(newState)) {
        return newState;
      }

      return null;
    }
  }]);

  return PopupSelect;
}(PureComponent);

_defineProperty(PopupSelect, "defaultProps", {
  closeMenuOnSelect: true,
  components: {},
  maxMenuHeight: 300,
  maxMenuWidth: 440,
  minMenuWidth: 220,
  popperProps: {},
  searchThreshold: 5,
  styles: {},
  options: []
});

export { PopupSelect as default };
