// @atlaskit/select/dist/cjs/Select.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.SelectWithoutAnalytics = void 0;

var _reactSelect = _interopRequireDefault(require("react-select"));

var _analyticsNext = require("@atlaskit/analytics-next");

var _createSelect = _interopRequireDefault(require("./createSelect"));

var packageName = "@atlaskit/select";
var packageVersion = "15.2.2";
var SelectWithoutAnalytics = (0, _createSelect.default)(_reactSelect.default);
exports.SelectWithoutAnalytics = SelectWithoutAnalytics;
var createAndFireEventOnAtlaskit = (0, _analyticsNext.createAndFireEvent)('atlaskit');

var _default = (0, _analyticsNext.withAnalyticsContext)({
  componentName: 'select',
  packageName: packageName,
  packageVersion: packageVersion
})((0, _analyticsNext.withAnalyticsEvents)({
  onChange: createAndFireEventOnAtlaskit({
    action: 'changed',
    actionSubject: 'option',
    attributes: {
      componentName: 'select',
      packageName: packageName,
      packageVersion: packageVersion
    }
  })
})(SelectWithoutAnalytics));

exports.default = _default;
