// @atlaskit/theme/dist/esm/components/hooks/use-theme-reset-styles.js
import { useEffect, useRef } from 'react';
import { heading, link, linkActive, linkHover, linkOutline, subtleHeading, subtleText, text } from '../../colors';
export var SELECTOR = 'old-ds-theme-mode';
var baseResetStyles = [{
  theme: {
    mode: 'light'
  }
}, {
  theme: {
    mode: 'dark'
  }
}].map(function (theme) {
  return "\n  .".concat(theme.theme.mode, "-").concat(SELECTOR, " {\n    color: ").concat(text(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " a {\n    color: ").concat(link(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " a:hover {\n    color: ").concat(linkHover(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " a:active {\n    color: ").concat(linkActive(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " a:focus {\n    outlineColor: ").concat(linkOutline(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h1,\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h2,\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h3,\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h4,\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h5 {\n    color: ").concat(heading(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " h6 {\n    color: ").concat(subtleHeading(theme), ";\n  }\n  .").concat(theme.theme.mode, "-").concat(SELECTOR, " small {\n    color: ").concat(subtleText(theme), ";\n  }");
}).join('\n');

var getStylesheetResetCSS = function getStylesheetResetCSS(backgroundColor) {
  return "body { background: ".concat(backgroundColor, "; }");
};

var UNIQUE_INTERNAL_ID = 'ds--theme--ak-theme-provider';
/**
 * This hooks conditionally sets body styles based on the theme mode applied.
 *
 * @param backgroundColor The background color to be applied at the root level of the application
 */

var useThemeResetStyles = function useThemeResetStyles(backgroundColor) {
  var stylesheet = useRef(null);
  useEffect(function () {
    var hasNode = document.getElementById(UNIQUE_INTERNAL_ID); // Bail out if the AKThemeProvider has already set the body
    // Child nodes should not take precedence over a root node setting body bg

    if (hasNode) {
      return;
    }

    stylesheet.current = document.createElement('style');

    if (document && document.head) {
      stylesheet.current.id = UNIQUE_INTERNAL_ID;
      var firstStyleElement = document.head.querySelector('style:not([data-styled-components],[data-emotion])'); // This is _likely_ to be the css reset, ideally this isn't required as we shouldn't
      // reset background color here, but it's still relied on by editor and mobile.

      if (firstStyleElement) {
        document.head.insertBefore(stylesheet.current, firstStyleElement.nextSibling);
      } else {
        document.head.prepend(stylesheet.current);
      }
    }

    return function () {
      if (stylesheet.current && document && document.head) {
        document.head.removeChild(stylesheet.current);
        stylesheet.current = null;
      }
    };
  }, []);
  useEffect(function () {
    if (stylesheet.current) {
      var css = getStylesheetResetCSS(backgroundColor);
      stylesheet.current.innerHTML = css + baseResetStyles;
    }
  }, [backgroundColor]);
};

export default useThemeResetStyles;
